sklearn/metrics/_classification.py

```diff
@@ -1633,6 +1633,183 @@ def precision_recall_fscore_support(
+            by support (the number of true instances for each label). This
+            alters 'macro' to account for label imbalance; it can result in an
+            F-score that is not between precision and recall.
+        ``'samples'``:
+            Calculate metrics for each instance, and find their average (only
+            meaningful for multilabel classification where this differs from
+            :func:`accuracy_score`).
+
+    sample_weight : array-like of shape (n_samples,), default=None
+        Sample weights.
+
+    Returns
+    -------
+    positive_likelihood_ratio : float (if average is not None) or array of float,\
+        shape = [n_unique_labels]
+        Positive likelihood ratio.
+
+    negative_likelihood_ratio : float (if average is not None) or array of float,\
+        shape = [n_unique_labels]
+        Negative likelihood ratio.
+
+    Notes
+    -----
+    When ``false positive == 0``, positive likelihood ratio is undefined.
+    When ``true negative == 0``, negative likelihood ratio is undefined.
+    When ``true positive + false negative == 0`` both ratios are undefined.
+    In such cases, ``UndefinedMetricWarning`` will be raised.
+
+    References
+    ----------
+    .. [1] `Wikipedia entry for the Likelihood ratios in diagnostic testing
+           <https://en.wikipedia.org/wiki/Likelihood_ratios_in_diagnostic_testing>`_.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> from sklearn.metrics import likelihood_ratios
+    >>> y_true = np.array(['cat', 'dog', 'pig', 'cat', 'dog', 'pig'])
+    >>> y_pred = np.array(['cat', 'pig', 'dog', 'cat', 'cat', 'dog'])
+    >>> likelihood_ratios(y_true, y_pred, average='macro')
+    (1.33..., 1.11...)
+    >>> likelihood_ratios(y_true, y_pred, average='micro')
+    (1.0, 1.0)
+    >>> likelihood_ratios(y_true, y_pred, average='weighted')
+    (1.33..., 1.11...)
+
+    It is possible to compute per-label likelihood ratios instead of averaging:
+
+    >>> likelihood_ratios(y_true, y_pred, average=None,
+    ... labels=['pig', 'dog', 'cat'])
+    (array([0., 0., 4.]), array([1.33..., 2.        , 0.        ]))
+    """
+
+    labels = _check_set_wise_labels(y_true, y_pred, average, labels, pos_label)
+
+    # Compute a confusion matrix for each class or sample
+    samplewise = average == "samples"
+    MCM = multilabel_confusion_matrix(
+        y_true,
+        y_pred,
+        sample_weight=sample_weight,
+        labels=labels,
+        samplewise=samplewise,
+    )
+    tp = MCM[:, 1, 1]
+    tn = MCM[:, 0, 0]
+    fp = MCM[:, 0, 1]
+    fn = MCM[:, 1, 0]
+
+    if average == "micro":
+        tp = np.array([tp.sum()])
+        tn = np.array([tn.sum()])
+        fp = np.array([fp.sum()])
+        fn = np.array([fn.sum()])
+
+    true_sum = tp + fn
+    false_sum = tn + fp
+    pos_num = tp * false_sum
+    pos_denom = fp * true_sum
+    neg_num = fn * false_sum
+    neg_denom = tn * true_sum
```

Comment:
Keeping the calculations in terms of `tp`, `fn`, `tn` and `fp` captures all the possibilities that lead to a zero division. For instance, if the negative likelihood ratio is calculated as `LR− = (1 − sensitivity) / specificity`, the only seemingly case of zero division is when `specificity = 0`, i.e., when `tn = 0`. Computing `LR− = [fn * (tn + fp)]/[tn * (tp + fn)]` we realize that `tp + fn = 0 || tn!=0` also leads to a zero division.

Reply:
On a second thought, this is highly non-trivial.

In the case of the positive likelihood ratio we have 2 possible zero divisions: 
- when `fp=0`, which has a real interpretation (the classifier being perfect) and should return `infty`;
- when `tp = fn = 0`, which means that no samples of the positive class were observed during training. This can happen when cross-validating highly imbalanced data. Such cases should return `nan` in order to be easily avoided when averaging over folds.

For the negative likelihood ratio the 2 possible zero divisions can be:
- when `tn=0`, which should not lead to `infty`, as a value of ``LR-`` greater than 1 is invalid beacause it would indicate an increase in the odds of a sample belonging to the positive class after being classified as negative, as if the act of classifying caused the condition.
- when `tp = fn = 0`, which means that no samples of the positive class were observed during training. This can also happen when cross-validating highly imbalanced data and such cases should return `nan`.

An additional problem arises when `fp = tn = 0`, which means that no samples of the negative class were observed during training, i.e. the symmetric problem when having highly imbalanced data. Such cases lead to a positive likelihood ratios equal to 0, which is possible yet invalid, as a value of ``LR+`` lower than 1 would indicate that the odds of a sample being a true positive decrease with respect to the pre-test odds.

All the above cases should rise a warning explaining the mentioned particularities.

Thoughts on this, @GaelVaroquaux ?

Reply:
Maybe the easiest thing to do is to return `np.nan` in all the cases above, but with warning messages that are descriptive enough of each case.